tests/std/tests/P2502R2_generator/test.cpp

```diff
@@ -0,0 +1,429 @@
+static_assert(static_checks<gen_traits<int&, int>, int, int&, int&&>());
+
+template <class Traits, class ValueType, class ReferenceType, class RvalueReferenceType, class R, class V, class A>
+void test_one(generator<R, V, A> g0, invocable<generator<R, V, A>> auto adaptor, ranges::input_range auto&& expected)
+    requires ranges::input_range<decltype(adaptor(move(g0)))>
+{
+    static_assert(same_as<generator<R, V, A>, typename Traits::generator>);
+    static_assert(static_checks<Traits, ValueType, ReferenceType, RvalueReferenceType>());
+
+    auto g1 = move(g0);
+    auto i  = ranges::cbegin(expected);
+    for (auto&& x : adaptor(move(g1))) {
+        assert(i != ranges::cend(expected));
+        assert(*i == x);
+        ++i;
+        // Verify iterator stays valid after move assignment
+        ranges::swap(g0, g1);
+    }
+    assert(i == ranges::cend(expected));
+}
+
+template <class Traits, class ValueType, class ReferenceType, class RvalueReferenceType, class R, class V, class A,
+    ranges::input_range Ex>
+void test_one(generator<R, V, A> g0, Ex&& expected) {
+    return test_one<Traits, ValueType, ReferenceType, RvalueReferenceType>(move(g0), identity{}, forward<Ex>(expected));
+}
+
+// Some simple end-to-end tests, mostly from the Working Draft or P2502R2
+generator<int> ints(int start = 0) {
+    while (true) {
+        co_yield start++;
+    }
+}
+
+template <ranges::input_range Rng1, ranges::input_range Rng2>
+generator<tuple<ranges::range_reference_t<Rng1>, ranges::range_reference_t<Rng2>>,
+    tuple<ranges::range_value_t<Rng1>, ranges::range_value_t<Rng2>>>
+    co_zip(Rng1 r1, Rng2 r2) {
+    auto it1        = ranges::begin(r1);
+    auto it2        = ranges::begin(r2);
+    const auto end1 = ranges::end(r1);
+    const auto end2 = ranges::end(r2);
+    for (; it1 != end1 && it2 != end2; ++it1, ++it2) {
+        co_yield {*it1, *it2};
+    }
+}
+
+void zip_example() {
+    using V = tuple<int, int>;
+    using R = tuple<int&, int&>;
+
+    auto g0 = co_zip(array{1, 2, 3}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{1, 10}, tuple{2, 20}, tuple{3, 30}});
+
+    g0 = co_zip(array{3, 2, 1}, vector{10, 20, 30, 40, 50});
+    test_one<gen_traits<R, V>, V, R, R>(move(g0), array{tuple{3, 10}, tuple{2, 20}, tuple{1, 30}});
+}
+
+template <class Reference = const int&>
+generator<Reference, int> co_upto(const int hi) {
+    assert(hi >= 0);
+    for (int i = 0; i < hi; ++i) {
+        co_yield i;
+    }
+}
+
+void test_weird_reference_types() {
+    constexpr int n = 32;
+    { // Test mutable lvalue reference type
+        auto r   = co_upto<int&>(n);
+        auto pos = r.begin();
+        for (int i = 0; i < n / 2; ++i, ++*pos, ++pos) {
+            assert(pos != r.end());
+            assert(*pos == 2 * i);
+        }
+        assert(pos == r.end());
+    }
+
+#if !(defined(__clang__) && defined(_M_IX86)) // TRANSITION, LLVM-56507
+    { // Test with mutable xvalue reference type

```

Comment:
"xvalue reference type" is not actually a thing.

Reply:
My brain cannon has a taxonomy of iterators mostly based on the value category of `*i`: lvalue iterators, xvalue iterators, pure prvalue iterators (reference type and value type are the same), and proxy iterators (value category of `*i` is a prvalue, but not `value_type`). These comments are abbreviations of "test with reference type chosen so the iterator produced is a meow iterator" which is great and meaningful if you happen to be me. Apparently not so good for not me.